### DISCUSSION TITLE
volt - v5.0.0RC3

### DISCUSSION THREAD

**Discussion by davidcorreia** [phalcon/cphalcon#16021] [General]
hi,
just tried to test RC3 but got an error ` PHP Parse error:  syntax error, unexpected identifier "fa", expecting "]" in /var/www/html/cache/_var_www_html_app_views_index.volt.php on line 632`

which is `<?= \Phalcon\Tag::linkTo(["user/profile", "<i class="fa fa-user fa-fw"></i> Perfil do utilizador"]) ?>`
which gives rise to the following
`{{ link_to('user/profile', '<i class="fa fa-user fa-fw"></i> Perfil do utilizador') }}`

with an image it's easy to see the problem
![image](https://user-images.githubusercontent.com/16776779/178353974-18e392d9-06da-45ba-80a7-6e261072bb2b.png)

with RC2 it's working

**Comment by borisdelev** [comment]
... is it work if you escape it?
```
{{ link_to('user/profile', '<i class="fa fa-user fa-fw"></i> Perfil do utilizador'|e) }}
```

**Reply by davidcorreia** [reply]
unfortunately it does not :(

**Reply by borisdelev** [reply]
I can confirm- have same issue :/ to fix that (before phalcon update)... i switch " with '... So you can try too:
```
{{ link_to('user/profile', "<i class='fa fa-user fa-fw'></i> Perfil do utilizador" ) }}
```

**Reply by davidcorreia** [reply]
that will probably work....
however... i think " should work as well :/

**Reply by borisdelev** [reply]
There is issue request. They will fix it.

**Reply by davidcorreia** [reply]
oh... i did not saw it. https://github.com/phalcon/cphalcon/issues/16019
thank you for mention it

**Comment by niden** [comment]
@davidcorreia this has been fixed in RC4 correct?

**Reply by davidcorreia** [reply]
@niden Thank you for warning me about RC4.

I tried to test however got some problems with the form to perform login.

`PHP Parse error:  syntax error, unexpected token "=>", expecting ")" `

`{{ form("login/index", "method": "post") }}`
originated this
` <?= $this->tag->form("login/index", 'method' => "post") ?>`
missing []


in the realease there is a meantion of renaming form

    Renamed Phalcon\Tag::form() to Phalcon\Tag::formLegacy (helper form_legacy) [#16019](https://github.com/phalcon/cphalcon/issues/16019)


I confirm the link_to is solved with ' or "
```
<li>{{ link_to('user/profile', '<i class="fa fa-user fa-fw"></i> Perfil do utilizador') }}</li>
<li>{{ link_to('user/profile', "<i class='fa fa-user fa-fw'></i> Perfil do utilizador") }}</li>

translate to

<li><?= \Phalcon\Tag::linkTo(['user/profile', '<i class="fa fa-user fa-fw"></i> Perfil do utilizador']) ?></li>
<li><?= \Phalcon\Tag::linkTo(['user/profile', '<i class=\'fa fa-user fa-fw\'></i> Perfil do utilizador']) ?></li>
```

**Comment by niden** [comment]
OK great about the `link_to`.

For the `form` if you want to use the existing code, just rename it to `form_legacy()`. This will keep the same functionality and will use internally the `Phalcon\Tag` component. If you use `form()` that will use the new component from the `TagFactory` and that has a different syntax.

If you have the time, you could change all the `form` elements to follow the new parameters etc. which will save you from a future upgrade since `Phalcon\Tag` will eventually be removed.

**Comment by davidcorreia** [comment]
sorry but i didn't quite understand.
if the `Phalcon\Tag` will be removed what should i use? should i stop using volt?

**Comment by niden** [comment]
`Phalcon\Tag` will be removed in the future. All of its functionality with some modifications of course is already in the `Phalcon\Html\TagFactory`.

You can use (for now) both in your code and also in Volt. The one clash that we had with names was `form()`, so we kept the `TagFactory::form()` as is (this is why you see the issue above) and renamed the volt helper to `form_legacy` for `Phalcon\Tag::formLegacy`

You can use the `Phalcon\Tag` component as before, you only have to rename in Volt calls to `form()` -> `form_legacy()`

Once the `Phalcon\Tag` is removed in a future version, then you will have to go back and change all that `form_legacy` to `form` and also adjust the method call.

I hope the above makes more sense :)

**Reply by s-ohnishi** [reply]
Since `Phalcon\Tag` is scheduled to be removed in the future, I try to write in `Phalcon\Html\TagFactory` as much as possible.
The opening tag of a form can be written as `{{ tag.form( ... ) }}`, but what about the closing tag?
`{{ end_form() }}` can only be used now, right?
Should I write raw HTML?

**Reply by niden** [reply]
@s-ohnishi You have two options

```twig
{{ end_form }}
```
will produce `</form>` and it is (currently) using `Phalcon\Tag`

You can also use:

```twig
{{ close('form') }}
```
or
```twig
{{ tag.close('form') }}
```

which will produce `</form>` but it uses the `Phalcon\Html\TagFactory`

I will change the `end_form()` to point to the `TagFactory` in the next day or so.

**Reply by s-ohnishi** [reply]
`tag.close()` was used for such purposes.
I didn't understand it at all.
I don't know the tag "close", but I thought it was just a lack of study.
Thank you very much.
